snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java

```diff
@@ -15,169 +15,847 @@
+		final QueryExpressionDiff diff = QueryExpressionDiff.create(
+			context, 
+			inclusions, 
+			exclusions, 
+			optimizedInclusions, 
+			optimizedExclusions);
+
+		log.info("Optimize resulted in {} addition(s) to inclusion clauses, {} addition(s) to exclusion clauses and {} removal(s)", 
+			diff.getAddToInclusion().size(),
+			diff.getAddToExclusion().size(),
+			diff.getRemove().size());
 		
-		// if there are no single concept inclusions to optimize, exit early
-		if (singleConceptInclusions.isEmpty()) {
+		if (diff.getAddToExclusion().isEmpty() && diff.getAddToExclusion().isEmpty() && diff.getRemove().isEmpty()) {
 			return QueryExpressionDiffs.EMPTY;
 		}
 		
-		// Record the ancestors (both direct and indirect) of each single concept inclusion
-		final Multimap<String, QueryExpression> membersByAncestor = HashMultimap.create();
-		
-		final int partitionSize = context.service(RepositoryConfiguration.class)
-			.getIndexConfiguration()
-			.getTermPartitionSize();
-		
-		Iterables.partition(singleConceptInclusions.keySet(), partitionSize).forEach(batchIds -> {
-			SnomedRequests.prepareSearchConcept()
-				.filterByIds(batchIds)
-				.setLimit(batchIds.size())
-				.build()
-				.execute(context)
-				.forEach(child -> {
-					final Collection<QueryExpression> childExpressions = singleConceptInclusions.get(child.getId());
-					final List<String> parentIds = child.getParentIdsAsString();
-					final List<String> ancestorIds = child.getAncestorIdsAsString();
-					
-					parentIds.forEach(parentId -> {
-						if (!IComponent.ROOT_ID.equals(parentId) && !Concepts.ROOT_CONCEPT.equals(parentId)) {
-							membersByAncestor.putAll(parentId, childExpressions);
-						}
-					});
-					
-					ancestorIds.forEach(ancestorId -> {
-						if (!IComponent.ROOT_ID.equals(ancestorId) && !Concepts.ROOT_CONCEPT.equals(ancestorId)) {
-							membersByAncestor.putAll(ancestorId, childExpressions);
-						}
-					});
-				});
-		});
+		return addLabelsToDiff(context, locales, diff);
+	}
+
+	private Set<String> optimizeInclusions(final BranchContext context, final Collection<QueryExpression> inclusions) {
+		optimizedInclusions = newArrayList();
+
+		// Pinned inclusions are checked for redundancy but otherwise should appear unmodified in optimized output 
+		final List<QueryExpression> pinnedInclusions = optimizePinned(context, inclusions);
+		log.info("Found {} unique pinned inclusion(s)", pinnedInclusions.size());
+		applyInclusions(context, pinnedInclusions);
+
+		// See if concepts still to be processed can be converted to "* : c1 = c2" expressions
+		final SnomedRelationshipStats inclusionRelationshipStats = SnomedRelationshipStats.create(
+			context, 
+			conceptsToInclude, 
+			relationshipSearchBySource, 
+			relationshipSearchByTypeAndDestination);
+
+		filterRefinementsForInclusion(inclusionRelationshipStats);
+		final List<QueryExpression> refinementInclusions = inclusionRelationshipStats.optimizeRefinements(context);
```

Comment:
This will always try to optimize using refinements first, right? Isn't that costly? I mean if there are 10k conceptIds then we always start with a relationship fetch according to this method call. Would it be possible to move this to a configuration value and check how long it takes to optimize a 1k and 10k refset with refinements enabled and disabled? Thanks!

Reply:
Runtimes are in milliseconds:

Scenario*      | 1k disabled | 1k enabled | 10k disabled | 10k enabled
-------------- | ------------ | ------------- | -------------- | -------------
Attempt 1     | 7820           | 8097            | 59672            | 63078
Attempt 2     | 7684           | 8029            | 61261            | 61407
Attempt 3     | 7700           | 7961            | 59968            | 62501
**Average**  | 7734.67      | 8029            | 60300.33       | 62328.67

<sup>*Actual member counts were 932 (VS 999029411000230108) and 9734 (VS 999021651000230109)</sup>